### PR TITLE
Container API first pass refactor and cleanup

### PR DESCRIPTION
**Pull request by StripedMonkey**
This is a refactor of the Container API to make it a little nicer.

I'm looking for feedback from @Bryntet above anything, but we'll see when we get there.

Once changes have been discussed, I will chunk out commits a little better.

### CODE REVIEW

**Comment by Bryntet** [comment]
I'm on a walk rn, I'll be able to take a look at this in like half an hour to an hour

File: pumpkin-inventory/src/lib.rs [hunk 1]
```diff
@@ -66,29 +67,36 @@ pub trait Container: Sync + Send {
...
     fn all_combinable_slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
-        self.all_slots()
+        self.iter_slots_mut().collect_vec()
     }
 
     fn internal_pumpkin_id(&self) -> u64 {
         0
     }
+
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>>;
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:96]: Maybe put a doc comment here, explaining the difference between the two `Option`

File: pumpkin-inventory/src/lib.rs [hunk 2]
```diff
@@ -212,26 +223,68 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>> {
+        todo!()
+    }
+
+    fn get_slot_mut(&mut self, slot: usize) -> Option<&mut Option<ItemStack>> {
+        todo!()
+    }
+
+    fn size(&self) -> usize {
+        todo!()
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:269]: this should be:
```rust
match &self.container {
    Some(container) => 36 + container.size(),
    None => self.inventory.size()
}
```

File: pumpkin/src/client/container.rs [hunk 3]
```diff
@@ -231,17 +227,17 @@ impl Player {
...
-                    } else {
-                        slots.skip(36).rev().find_map(find_condition)
-                    };
-                    if let Some(slot) = slots {
-                        let mut item_slot = container.all_slots()[slot].map(|i| i.to_owned());
-                        container.handle_item_change(&mut item_slot, slot, MouseClick::Left)?;
-                        *container.all_slots()[slot] = item_slot;
-                    }
+                    let all_slots = container.iter_slots_mut().enumerate();
+                    // let slots = if slot > 35 {
```
Comment (Bryntet) [inline pumpkin/src/client/container.rs:231]: this code works, but only for the case of being inside of Inventory. Implementing this should be easy now if we use `Container::size`

File: pumpkin-inventory/src/player.rs [hunk 4]
```diff
@@ -57,13 +57,19 @@ impl PlayerInventory {
             if !(0..=45).contains(&slot) {
```
Comment (Bryntet) [inline pumpkin-inventory/src/player.rs:57]: this can be removed, as you added a check below

File: pumpkin-inventory/src/player.rs [hunk 5]
```diff
@@ -57,13 +57,19 @@ impl PlayerInventory {
...
                 Err(InventoryError::InvalidSlot)?
             }
-            *self.all_slots()[slot] = item;
+            let Some(slot) = self.get_slot_mut(slot) else {
```
Comment (Bryntet) [inline pumpkin-inventory/src/player.rs:60]: change this to `let slot = self.get_slot_mut(slot).ok_or(Err(InventoryError::InvalidSlot))?;`

File: pumpkin-inventory/src/player.rs [hunk 6]
```diff
@@ -57,13 +57,19 @@ impl PlayerInventory {
...
+                return Err(InventoryError::InvalidSlot);
+            };
+            *slot = item;
             return Ok(());
         }
         let slot_condition = self.slot_condition(slot)?;
         if let Some(item) = item {
             if slot_condition(&item) {
-                *self.all_slots()[slot] = Some(item);
+                let Some(slot) = self.get_slot_mut(slot) else {
```
Comment (Bryntet) [inline pumpkin-inventory/src/player.rs:69]: change this to `let slot = self.get_slot_mut(slot).ok_or(Err(InventoryError::InvalidSlot))?;`

File: pumpkin-inventory/src/player.rs [hunk 7]
```diff
@@ -156,19 +150,55 @@ impl Container for PlayerInventory {
...
+            1..=4 => Some(&mut self.crafting[slot - 1]),
+            5..=8 => Some(&mut self.armor[slot - 5]),
+            9..=44 => Some(&mut self.items[slot - 9]),
+            45 => Some(&mut self.offhand),
+            _ => None,
+        }
+    }
+
+    fn size(&self) -> usize {
+        todo!()
```
Comment (Bryntet) [inline pumpkin-inventory/src/player.rs:202]: this is 46 i think

File: pumpkin-inventory/src/lib.rs [hunk 8]
```diff
@@ -212,26 +223,68 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
-                slots
+                let container_size = container.size();
+
+                Box::new(ExactSizedChain {
+                    iter: Box::new(
+                        container
+                            .iter_slots_mut()
+                            .chain(self.inventory.all_combinable_slots_mut()),
+                    ),
+                    len: 36 + container_size,
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:239]: maybe should be self.size()

File: pumpkin-inventory/src/lib.rs [hunk 9]
```diff
@@ -212,26 +223,68 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
+        &'s mut self,
+    ) -> Box<(dyn ExactSizeIterator<Item = &mut Option<ItemStack>> + 's)> {
+        match &mut self.container {
             Some(container) => {
-                let mut slots = container.all_slots();
-                slots.extend(self.inventory.all_combinable_slots_mut());
-                slots
+                let container_size = container.size();
+
+                Box::new(ExactSizedChain {
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:233]: I think you could extract out this to a function that does the same thing, cause this code appears in like two places rn

File: pumpkin-inventory/src/open_container.rs [hunk 10]
```diff
@@ -62,11 +62,27 @@ impl Container for Chest {
...
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>> {
+        todo!()
+    }
+
+    fn get_slot_mut(&mut self, slot: usize) -> Option<&mut Option<ItemStack>> {
+        todo!()
+    }
+
+    fn size(&self) -> usize {
+        todo!()
```
Comment (Bryntet) [inline pumpkin-inventory/src/open_container.rs:86]: this is 27

**Review by Bryntet — COMMENTED** [review]
Very nice PR, still a few todo!():s that need to be sorted out, and a few places where you can shorten what's being written. but overall very good, all the structural changes have my full approval!

File: pumpkin-inventory/src/lib.rs [hunk 11]
```diff
@@ -206,32 +215,51 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
+        container.window_name()
     }
 
-    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>> {
+    fn iter_slots_mut<'s>(&'s mut self) -> Box<(dyn Iterator<Item = &mut Option<ItemStack>> + 's)> {
+        match &mut self.container {
+            Some(container) => Box::new(
+                container
+                    .iter_slots_mut()
+                    .chain(self.inventory.iter_slots_mut()),
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:229]: If container is `Some` on `OptionallyCombinedContainer` only chain it with the main inventory slots on `PlayerInventory`. The client is not able to change things (like armour) while inside of a container. 

See https://wiki.vg/Inventory for a nice visualisation of where each slot goes, depending on which container that is opened.

File: pumpkin-inventory/src/lib.rs [hunk 12]
```diff
@@ -206,32 +215,51 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
+            None => self.inventory.iter_slots(),
         }
     }
+
+    fn get_slot(&self, slot: usize) -> Option<&Option<ItemStack>> {
+        if (0..self.inventory.size()).contains(&slot) {
+            return self.inventory.get_slot(slot);
+        }
+        if let Some(container) = &self.container {
+            return container.get_slot(slot);
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:247]: you need to `return container.get_slot(slot-self.inventory.size())`

File: pumpkin-inventory/src/lib.rs [hunk 13]
```diff
@@ -206,32 +215,51 @@ impl<'a> Container for OptionallyCombinedContainer<'a, 'a> {
...
+        }
+        None
+    }
+
+    fn get_slot_mut(&mut self, slot: usize) -> Option<&mut Option<ItemStack>> {
+        if (0..self.inventory.size()).contains(&slot) {
+            return self.inventory.get_slot_mut(slot);
+        }
+        if let Some(container) = &mut self.container {
+            return container.get_slot_mut(slot);
```
Comment (Bryntet) [inline pumpkin-inventory/src/lib.rs:257]: you need to `return container.get_slot_mut(slot-self.inventory.size())`

File: pumpkin-inventory/src/lib.rs (hunk 12 above)
Comment (Bryntet) [reply pumpkin-inventory/src/lib.rs:247]: actually nevermind, this should be a match statement, looking something like:
```rust
match &self.container {
    Some(container) => {
        if (0..container.size()).contains(&slot) {
            container.get_slot(slot)
        } else {
            self.inventory.items.get(slot)
        }
    None => self.inventory.get_slot(slot)
}
```

see https://wiki.vg/Inventory#Chest

File: pumpkin-inventory/src/lib.rs (hunk 13 above)
Comment (Bryntet) [reply pumpkin-inventory/src/lib.rs:257]: see amendment to comment above!

**Comment by StripedMonkey** [comment]
I'd like some, 

I've been working on other things since this PR. I'll give you access if you don't already this afternoon.

**Comment by Snowiiii** [comment]
I will close this for now, Feel free to reopen. I think that there are already so many conflicts that it would be easier to Start from scratch again

**Comment by Bryntet** [comment]
I'd like this not to be closed, I'll solve the merge conflicts.

**Pull request closed without merging**